File: typescript/end-to-end-applications/ai-image-workflows/src/transformer_service.ts
import * as restate from "@restatedev/restate-sdk";
import {WorkflowStep} from "./types/types";
import Jimp from "jimp";
import {TerminalError} from "@restatedev/restate-sdk";

export const transformerService = restate.service({
    name: "transformer",
    handlers: {
        rotate: async (ctx: restate.Context, wfStep: WorkflowStep) => {
            const { angle } = wfStep.parameters as { angle: number };
            ctx.console.info(`Rotating image with angle: ${angle}`);

            await ctx.run(async () => {
                const image = await getImage(wfStep.imgInputPath!);
                await image.rotate(angle).writeAsync(wfStep.imgOutputPath!);
            });

            return { msg: `[Rotated image with angle: ${angle}]` };
        },

        blur: async (ctx: restate.Context, wfStep: WorkflowStep) => {
            const { blur } = wfStep.parameters as { blur: number };
            ctx.console.info(`Blurring image with parameter ${blur}`);

            await ctx.run(async () => {
                const image = await getImage(wfStep.imgInputPath!);
                await image.blur(blur).writeAsync(wfStep.imgOutputPath!);
            });

            return {
                msg: `[Blurred image with strength param ${blur}]`
            };
        }
    }
})


function getImage(inputPath: string): Promise<Jimp> {
    try {
        return Jimp.read(inputPath);
    } catch (err) {
        throw new TerminalError("Error reading image: " + err)
    }
}

File: java/end-to-end-applications/workflow-interpreter/src/main/java/com/example/restatestarter/TransformerService.java
package com.example.restatestarter;

import com.example.restatestarter.types.WorkflowStep;
import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.springboot.RestateComponent;

import java.awt.image.BufferedImage;

import static com.example.restatestarter.utils.TransformerUtils.*;

@RestateComponent
@Service
public class TransformerService {

  @Handler
  public String rotate(WorkflowStep wfStep) throws Exception {
    int angle = (int) wfStep.parameters().get("angle");
    System.out.println("Rotating image with angle: " + angle);

    Restate.run("rotate", () -> {
      BufferedImage image = getImage(wfStep.imgInputPath());
      BufferedImage rotatedImage = rotateImage(image, angle);
      writeImage(rotatedImage, wfStep.imgOutputPath());
    });

    return "[Rotated image with angle: " + angle + "]";
  }

  @Handler
  public String blur(WorkflowStep wfStep) throws Exception {
    int blur = (int) wfStep.parameters().get("blur");
    System.out.println("Blurring image with parameter " + blur);

    Restate.run("blur", () -> {
      BufferedImage image = getImage(wfStep.imgInputPath());
      BufferedImage blurredImage = blurImage(image, blur);
      writeImage(blurredImage, wfStep.imgOutputPath());
    });

    return "[Blurred image with strength param " + blur + "]";
  }
}
